lib/ast.js

```diff
@@ -91,7 +91,15 @@ var AST_Token = DEFNODE("Token", "type value line col pos endline endcol endpos
 }, null);
 
 var AST_Node = DEFNODE("Node", "start end", {
-    clone: function() {
+    clone: function(deep) {
+        if (deep) {
+            var self = this.clone();
+            return self.transform(new TreeTransformer(function(node) {
+                if (node !== self) {
+                    return node.clone(true);
```

Comment:
Verified that this PR won't suffer the same cyclic issue as #769 due to the fact that those properties aren't child nodes and `TreeTransformer` won't walk over them.

And also, in those instances having it point to the same instance of `AST_Scope` rather than cloning it is the right thing to do.